### PR TITLE
BNH-205 added without guarantor to matching and filtering

### PR DESCRIPTION
**Pull request by TomHanmer29**
added an extra database field, filtering button, and extra filters on the matching data.
image
image

### CODE REVIEW

File: BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs [hunk 1]
```diff
@@ -34,7 +34,7 @@ public async void CheckIfImportWorks_CalledOnFileWithMissingColumn_ReturnsDanger
         await using var context = Fixture.CreateReadContext();
         var service = new CsvImportService(context, PostcodeService, Logger);
 
-        var data = Encoding.UTF8.GetBytes("Name,Phone,Postcode,Type,HasPet,NotInEET,UniversalCredit,HousingBenefits,Under35");
+        var data = Encoding.UTF8.GetBytes("Name,Phone,Postcode,Type,HasPet,NotInEET,UniversalCredit,HousingBenefits,Under35.NoGuarantor");
```
Comment (emilypatsko) [inline BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs:37]: Should that be a comma instead of a full stop?

Comment (oskarwilliams378) [inline BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs:37]: should this be 
```suggestion
        var data = Encoding.UTF8.GetBytes("Name,Phone,Postcode,Type,HasPet,NotInEET,UniversalCredit,HousingBenefits,Under35,NoGuarantor");

```

**Review by oskarwilliams378 — CHANGES_REQUESTED** [review]

File: web/Views/Admin/_MatchListTable.cshtml [hunk 2]
```diff
@@ -62,6 +62,10 @@
                         <i class="bi me-1 @GetPreferenceIcon(!tenant.NotInEET)"></i>
                         In EET (Employment, Education, or Training)
                     </div>
+                    <div class="@GetPreferenceFont(tenant.NoGuarantor)">
+                        <i class="bi me-1 @GetPreferenceIcon(tenant.NoGuarantor)"></i>
+                        Without rent guarantor
```
Comment (emilypatsko) [inline web/Views/Admin/_MatchListTable.cshtml:67]: Can we make this 'Has rent guarantor' instead? I think that's the default assumption, and the double negative (⛔ Without rent guarantor) is confusing

File: web/Views/Admin/_TenantTable.cshtml [hunk 3]
```diff
@@ -61,6 +61,10 @@
                         <i class="bi me-1 @GetPreferenceIcon(!tenant.NotInEET)"></i>
                         In EET (Employment, Education, or Training)
                     </div>
+                    <div class="@GetPreferenceFont(tenant.NoGuarantor)">
+                        <i class="bi me-1 @GetPreferenceIcon(tenant.NoGuarantor)"></i>
+                        Without rent guarantor
```
Comment (emilypatsko) [inline web/Views/Admin/_TenantTable.cshtml:66]: Same here

File: web/Database/TenantDbModel.cs [hunk 4]
```diff
@@ -19,4 +19,5 @@ public class TenantDbModel
     public bool? UniversalCredit { get; set; }
     public bool? HousingBenefits { get; set; }
     public bool? Under35 { get; set; }
+    public bool? NoGuarantor { get; set; }
```
Comment (emilypatsko) [inline web/Database/TenantDbModel.cs:22]: I think this should be 'HasGuarantor' - basically flipping the bool on its head. I know we'd like to make it so all the bools default to false, but I think that's just confusing for some of them (especially when it results in a double negative)

File: web/Views/Property/PropertyInputForm/TenantPreferences.cshtml [hunk 5]
```diff
@@ -45,7 +45,7 @@
     <div class="form-check">
         <input type="checkbox" asp-for="AcceptsBenefits" class="form-check-input"/>
         <label asp-for="AcceptsBenefits" class="form-check-label">
-            on Universal Credit or housing benefits?
+            on Housing Benefits?
```
Comment (emilypatsko) [inline web/Views/Property/PropertyInputForm/TenantPreferences.cshtml:48]: What's the reason for this change?

File: web/Views/Shared/Details/_TenantPreferences.cshtml [hunk 6]
```diff
@@ -35,7 +35,7 @@
         {
             <div class="@GetPreferenceFont(Model.Property.LandlordRequirements.AcceptsUnder35)">
                 <i class="bi me-1 @GetPreferenceIcon(Model.Property.LandlordRequirements.AcceptsUnder35)"></i>
-                Is over 35
+                Is under 35
```
Comment (emilypatsko) [inline web/Views/Shared/Details/_TenantPreferences.cshtml:38]: What's the reason for this change? Is this one that was missed out before?

**Review by emilypatsko — CHANGES_REQUESTED** [review]

File: BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs (hunk 1 above)
Comment (emilypatsko) [reply BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs:37]: I think @oskarwilliams378 beat me to it...

**Comment by emilypatsko** [comment]
Also, could we make it so that when viewing a property (including on the matching page), the requisite _for the property_ is shown as 

👍🏻 Accepts without guarantor
👎🏻 Accepts without guarantor

(obviously not with thumbs up/down but I couldn't find tick/cross emojis)

File: web/Views/Shared/Details/_TenantPreferences.cshtml (hunk 6 above)
Comment (TomHanmer29) [reply web/Views/Shared/Details/_TenantPreferences.cshtml:38]: Yeah this was missed out I think. Its different on the other property pages

File: web/Views/Property/PropertyInputForm/TenantPreferences.cshtml (hunk 5 above)
Comment (TomHanmer29) [reply web/Views/Property/PropertyInputForm/TenantPreferences.cshtml:48]: The line above says "on universal credit" and affects acceptsCredit, it makes no sense to have the on universal credit bit here any more

File: BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs [hunk 7]
```diff
@@ -26,6 +24,8 @@ public CsvImportServiceTests(TestDatabaseFixture fixture)
         MessageHandler = A.Fake<HttpMessageHandler>();
         HttpClient = new HttpClient(MessageHandler);
     }
+    private TestDatabaseFixture Fixture { get; }
+    private ILogger<CsvImportService> Logger { get; }
```
Comment (emilypatsko) [inline BricksAndHearts.UnitTests/ServiceTests/CsvImport/CsvImportServiceTests.cs:28]: Is there a reason these have moved down here?

File: web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public partial class AddHasGuarantorAndChangeEETInTenants : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.RenameColumn(
+                name: "NotInEET",
+                table: "Tenant",
+                newName: "InEET");
+
```
Comment (emilypatsko) [inline web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:15]: I think technically to preserve the 'trueness' of our data, we'll want to negate any existing data in the column by adding a line to execute some SQL, like:
```
Sql("UPDATE Tenant SET NotInEET = 1- NotInEET;");
```
@oskarwilliams378 does this look right to you?

Comment (emilypatsko) [reply web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:15]: And we'd need to do the same again in the `Down` method below

File: web/Services/TenantService.cs [hunk 9]
```diff
@@ -64,12 +64,14 @@ private IQueryable<TenantDbModel> GetFilteredTenantQuery(HousingRequirementModel
...
-                                          && (filters.AcceptsNotInEET == null || t.NotInEET == filters.AcceptsNotInEET)
+                                          && (filters.AcceptsNotInEET == null || t.InEET == !filters.AcceptsNotInEET)
                                           && (filters.AcceptsCredit == null
                                               || t.UniversalCredit == filters.AcceptsCredit)
                                           && (filters.AcceptsBenefits == null
                                               || t.HousingBenefits == filters.AcceptsBenefits)
-                                          && (filters.AcceptsUnder35 == null || t.Under35 == filters.AcceptsUnder35));
+                                          && (filters.AcceptsUnder35 == null || t.Under35 == filters.AcceptsUnder35)
+                                          && (filters.AcceptsWithoutGuarantor == null
+                                              || t.HasGuarantor == filters.AcceptsWithoutGuarantor));
```
Comment (emilypatsko) [inline web/Services/TenantService.cs:74]: I think that should be `t.HasGuarantor == !filters.AcceptsWithoutGuarantor`, shouldn't it?
e.g. if `HasGuarantor` is false, the tenant doesn't have a guarantor, so we'd want `AcceptsWithoutGuarantor` to be true

File: web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+#nullable disable
+
+namespace BricksAndHearts.Migrations
+{
+    public partial class AddHasGuarantorAndChangeEETInTenants : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.RenameColumn(
```
Comment (oskarwilliams378) [inline web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:11]: You can add something into the migration here to make the db invert all existing true and false values :)

File: web/Services/TenantService.cs (hunk 9 above)
Comment (oskarwilliams378) [inline web/Services/TenantService.cs:74]: should this be inverted

File: web/Views/Admin/_MatchListTable.cshtml [hunk 11]
```diff
@@ -58,22 +58,26 @@
...
-                {
-                    propertyLink = Model.LandlordAndProperty.Property!.PublicViewLink,
-                    tenantEmail = tenant.Email,
-                    currentPropertyId = Model.LandlordAndProperty.Property.PropertyId
-                },
-                    FormMethod.Post))
-                {
-                    <button type="submit" class="btn btn-thin btn-outline-primary mx-3">Match</button>
-                }
+                    @using (Html.BeginForm("SendMatchLinkEmail", "Tenant", new
```
Comment (oskarwilliams378) [inline web/Views/Admin/_MatchListTable.cshtml:71]: I'd un indent this

File: web/Views/Property/PropertyInputForm/PropertyAccepts.cshtml [hunk 12]
```diff
@@ -45,7 +45,7 @@
     <div class="form-check">
         <input type="checkbox" asp-for="AcceptsBenefits" class="form-check-input"/>
         <label asp-for="AcceptsBenefits" class="form-check-label">
-            on Universal Credit or housing benefits?
+            on Housing Benefits?
```
Comment (emilypatsko) [inline web/Views/Property/PropertyInputForm/PropertyAccepts.cshtml:48]: Could you do me a favour and add a question mark to the one above (`on Universal Credit?`)

File: web/wwwroot/TenantImportCSV_Template.csv [hunk 13]
```diff
@@ -1,2 +1,2 @@
-﻿Name,Email,Phone,Postcode,Type,HasPet,NotInEET,UniversalCredit,HousingBenefits,Under35
-Example Person,persona@gmail.com,01304 676838,cb39aj,Single,TRUE,TRUE,TRUE,TRUE,FALSE
+﻿Name,Email,Phone,Postcode,Type,HasPet,InEET,UniversalCredit,HousingBenefits,Under35,HasGuarantor
+Example Person,persona@gmail.com,01304 676838,cb39aj,Single,TRUE,TRUE,TRUE,TRUE,FALSE,TRUE
```
Comment (oskarwilliams378) [inline web/wwwroot/TenantImportCSV_Template.csv:2]: Do you want to flip the InEET True False?

**Review by oskarwilliams378 — CHANGES_REQUESTED** [review]

File: web/Views/Shared/Details/_PropertyAccepts.cshtml [hunk 14]
```diff
@@ -1,58 +1,58 @@
...
+                <i class="bi me-1 @GetPreferenceIcon(Model.Property.LandlordRequirements.AcceptsBenefits)"></i>
+                On housing benefits
+            </div>
+            <div class="@GetPreferenceFont(Model.Property.LandlordRequirements.AcceptsNotInEET)">
+                <i class="bi me-1 @GetPreferenceIcon(Model.Property.LandlordRequirements.AcceptsNotInEET)"></i>
+                Not in EET (Employment, Education, or Training)
+            </div>
+            <div class="@GetPreferenceFont(Model.Property.LandlordRequirements.AcceptsWithoutGuarantor)">
+                <i class="bi me-1 @GetPreferenceIcon(Model.Property.LandlordRequirements.AcceptsWithoutGuarantor)"></i>
+                Accepts without guarantor
```
Comment (emilypatsko) [inline web/Views/Shared/Details/_PropertyAccepts.cshtml:54]: I think this should just say 'without guarantor', because these all come below the 'Property accepts' heading, and it would read weirdly if 'accepts' was there twice :)

**Review by emilypatsko — CHANGES_REQUESTED** [review]

File: web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs (hunk 8 above)
Comment (oskarwilliams378) [reply web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:15]: I'd possibly not do it this way (mathematically), but do it with
```UPDATE Tenant SET InEET = NOT InEET```
I'd also do it with the new column names after we update the column name

File: web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs (hunk 10 above)
Comment (emilypatsko) [reply web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:11]: Moving my comment from below: 
I think technically to preserve the 'trueness' of our data, we'll want to negate any existing data in the column by adding a line to execute some SQL, like:
```
Sql("UPDATE Tenant SET NotInEET = 1- NotInEET;");
```
@oskarwilliams378 does this look right to you?

Comment (emilypatsko) [reply web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:11]: (and we'd want to do the same in the `Down` method)

File: web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs (hunk 8 above)
Comment (emilypatsko) [reply web/Migrations/20220825144120_AddHasGuarantorAndChangeEETInTenants.cs:15]: Two things:
- I meant to use the new column name but had tested my SQL locally first and hadn't renamed the column, forgot to change it in my comment oops
- I _wanted_ to do it this way, but MSSQL didn't like it :( and I'm not sure why. I've updated my above comment to what I think it should be @oskarwilliams378

File: web/Views/Admin/TenantMatchList.cshtml [hunk 15]
```diff
@@ -4,5 +4,7 @@
     ViewData["Title"] = "Tenant List";
 }
 
+<h1 class="font-h1">Selected property</h1>
 <partial name="~/Views/Admin/_MatchListPropertyViewPartial.cshtml" model="@Model.LandlordAndProperty"/>
+<h1 class="font-h1">Matching tenants</h1>
```
Comment (emilypatsko) [inline web/Views/Admin/TenantMatchList.cshtml:9]: My bad, I should have mentioned this - you should only ever have one `<h1>` on a page! If you want them to both _look_ like `<h1>`s, you should apply the appropriate bootstrap class to them both :)

File: web/Views/Shared/Details/_PropertyAccepts.cshtml [hunk 16]
```diff
@@ -35,23 +35,23 @@
...
             <div class="@GetPreferenceFont(Model.Property.LandlordRequirements.AcceptsCredit)">
                 <i class="bi me-1 @GetPreferenceIcon(Model.Property.LandlordRequirements.AcceptsCredit)"></i>
                 On Universal Credit
             </div>
-            <div class="@GetPreferenceFont(!Model.Property.LandlordRequirements.AcceptsNotInEET)">
-                <i class="bi me-1 @GetPreferenceIcon(!Model.Property.LandlordRequirements.AcceptsNotInEET)"></i>
-                In EET (Employment, Education, or Training)
+            <div class="@GetPreferenceFont(Model.Property.LandlordRequirements.AcceptsBenefits)">
+                <i class="bi me-1 @GetPreferenceIcon(Model.Property.LandlordRequirements.AcceptsBenefits)"></i>
+                On Housing Benefits
```
Comment (emilypatsko) [inline web/Views/Shared/Details/_PropertyAccepts.cshtml:46]: "On Housing Benefits" doesn't need capitalising :) just "On housing benefits"

**Review by emilypatsko — CHANGES_REQUESTED** [review]
Basically done, just 2 v small comments

**Review by emilypatsko — APPROVED** [review]
